### ISSUE TITLE
Making X * b go fast for large sparse X

### ISSUE THREAD

**Issue by sakrejda** [stan-dev/stan#1341]
After much prodding from Bob, this is an issue summarizing a conversation strewn over a few stan-users exchanges...

Since Stan programs benefit from pre-computing as much as possible outside of the parameters/transformed parameters blocks, models can often be simplified and sped up by generating a model matrix and reducing the in-Stan computation to $X \times b$-type matrix multiplication.  This is applicable to a wide variety of hierarchical models as well as spline models with truncated basis functions (my current use case).  Stan does not yet have sparse data types, but it is possible to represent a sparse X matrix using a set of vectors and (for example) Eigen's compressed sparse column format.  I implemented a simple Stan function for doing this multiplication (https://groups.google.com/d/msg/stan-users/mtF3udlP3PM/nCLhXXBTlNIJ) and am in the process of re-doing an implementation that relies on Eigen's format as described at (http://eigen.tuxfamily.org/dox/group__TutorialSparse.html).  The new implementation with more explicit documentation is here: https://github.com/sakrejda/sparse-stan

From some informal tests I see that you can get all of the expected memory benefit and a small speed gain from a Stan function ignoring the gradient calculation. The gradient is simple and everybody says that's where the real speed gains are.  

To-do's:
- [x] finish implementing Stan-level function
- [x] find a similar C++ function to work from
- [x] follow the instructions for adding a C++ level function (stan-users and the static page).
- [ ] custom gradients??

**Comment by bob-carpenter** [comment]
> On Mar 5, 2015, at 1:54 PM, Krzysztof Sakrejda <notifications@github.com> wrote:
> 
> After much prodding from Bob,

The prodding was to have the discussion on stan-dev.

> this is an issue summarizing a conversation strewn over a few stan-users exchanges...
> 
> Since Stan programs benefit from pre-computing as much as possible outside of the parameters/transformed parameters blocks, models can often be simplified and sped up by generating a model matrix and reducing the in-Stan computation to $X \times b$-type matrix multiplication. 

Is there an example of where this kind of thing is faster?  My intuition
is that it'd be slower than just doing the custom multiplication in
the model.  

The advantage of (X * beta) is that we can do custom gradients.  
We should be doing them, but I'm not sure they're plugged
in everywhere, hence issue #1339.

> This is applicable to a wide variety of hierarchical models as well as spline models with truncated basis functions (my current use case). Stan does not yet have sparse data types, but it is possible to represent a sparse X matrix using a set of vectors and (for example) Eigen's compressed sparse column format. I implemented a simple Stan function for doing this multiplication (https://groups.google.com/d/msg/stan-users/mtF3udlP3PM/nCLhXXBTlNIJ) and am in the process of re-doing an implementation that relies on Eigen's format as described at (http://e igen.tux family.org/dox/group__TutorialSparse.html). The new implementation with more explicit documentation is here: https://github.com/sakrejda/sparse-stan
> 
> From some informal tests I see that you can get all of the expected memory benefit and a small speed gain from a Stan function ignoring the gradient calculation. The gradient is simple and everybody says that's where the real speed gains are.

Indeed. 

- Bob

**Comment by sakrejda** [comment]
On Mar 5, 2015 12:59 AM, "Bob Carpenter" <notifications@github.com> wrote:
>
>
> > On Mar 5, 2015, at 1:54 PM, Krzysztof Sakrejda <notifications@github.com>
wrote:
> >
> > After much prodding from Bob,
>
> The prodding was to have the discussion on stan-dev.
>

I don't think I can post on stan-dev, but somebody asked to make it an
issue... Doesn't matter now I guess, but I'm happy to have the conversation
on stan-dev

> > this is an issue summarizing a conversation strewn over a few
stan-users exchanges...
> >
> > Since Stan programs benefit from pre-computing as much as possible
outside of the parameters/transformed parameters blocks, models can often
be simplified and sped up by generating a model matrix and reducing the
in-Stan computation to $X \times b$-type matrix multiplication.
>
> Is there an example of where this kind of thing is faster? My intuition
> is that it'd be slower than just doing the custom multiplication in
> the model.

Now that you say that I can see why---in Stan you're doing only the
required calculations but with X*b there are inevitably some extra zero
entries.  The one case I have is doing the spline calculations.  I profiled
that one and everything looked good except the spline function was taking
up a good chunk of the time.  I redid the model to precalculate the weights
and it did go faster (you know, a few changes and it's one day instead of
two). I never went back do a formal comparison.

The bigger reason for using a model matrix is that it simplified the Stan
code a lot and made it possible to test the splines and predict much more
easily outside of Stan.  At that point including the sparse multiplication
is just a way of bringing the computational cost back down, and the memory.

>
> The advantage of (X * beta) is that we can do custom gradients.
> We should be doing them, but I'm not sure they're plugged in everywhere,
hence issue #1339.

Ok, but doing the sparse multiplication/gradients should still make sense,
even if that issue gets done.

Krzysztof

>
> > This is applicable to a wide variety of hierarchical models as well as
spline models with truncated basis functions (my current use case). Stan
does not yet have sparse data types, but it is possible to represent a
sparse X matrix using a set of vectors and (for example) Eigen's compressed
sparse column format. I implemented a simple Stan function for doing this
multiplication (
https://groups.google.com/d/msg/stan-users/mtF3udlP3PM/nCLhXXBTlNIJ) and am
in the process of re-doing an implementation that relies on Eigen's format
as described at (http://e igen.tux family.org/dox/group__TutorialSparse.html).
The new implementation with more explicit documentation is here:
https://github.com/sakrejda/sparse-stan
> >
> > From some informal tests I see that you can get all of the expected
memory benefit and a small speed gain from a Stan function ignoring the
gradient calculation. The gradient is simple and everybody says that's
where the real speed gains are.
